### ISSUE TITLE
import torch_xla_py.data_parallel as dp issues

### ISSUE THREAD

**Issue by williamFalcon** [pytorch/xla#659]
Hi, after following the install instructions, I get the following error when running tests. It seems to be caused by:
```
import torch_xla_py.data_parallel as dp issues
```

```
(tpu) waf2107@waf2107:~/pytorch/xla$ python app.py 
2019-05-07 01:40:15.424331: I tensorflow/compiler/xla/xla_client/computation_client.cc:125] 
Loading XRT configuration from /home/waf2107/.pytorch_tpu.conf
2019-05-07 01:40:15.424560: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device CPU:0 -> /job:tpu_worker/replica:0/task:0/device:XLA_CPU:0
2019-05-07 01:40:15.424578: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:0 -> /job:tpu_worker/replica:0/task:0/device:TPU:0
2019-05-07 01:40:15.424602: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:1 -> /job:tpu_worker/replica:0/task:0/device:TPU:1
2019-05-07 01:40:15.424611: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:2 -> /job:tpu_worker/replica:0/task:0/device:TPU:2
2019-05-07 01:40:15.424621: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:3 -> /job:tpu_worker/replica:0/task:0/device:TPU:3
2019-05-07 01:40:15.424630: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:4 -> /job:tpu_worker/replica:0/task:0/device:TPU:4
2019-05-07 01:40:15.424639: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:5 -> /job:tpu_worker/replica:0/task:0/device:TPU:5
2019-05-07 01:40:15.424647: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:6 -> /job:tpu_worker/replica:0/task:0/device:TPU:6
2019-05-07 01:40:15.424656: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
4] XRT device TPU:7 -> /job:tpu_worker/replica:0/task:0/device:TPU:7
2019-05-07 01:40:15.424665: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:4
7] XRT default device: TPU:0
2019-05-07 01:40:15.424701: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:8
77] Initializing TPU system for worker tpu_worker:0 at grpc://10.240.1.2:8470
2019-05-07 01:40:15.494162: E tensorflow/compiler/xla/xla_client/tf_logging.cc:11] Check fai
led: session->session()->Run({tensorflow::Output(result, 0)}, &outputs) == ::tensorflow::Sta
tus::OK() (Not found: Op type not registered 'XRTExecuteChained' in binary running on n-d9dc
d0ea-w-0. Make sure the Op and Kernel are registered in the binary running in this process. 
```

**Comment by dlibenzi** [comment]
The XRTExecuteChained API was recently added and will be available on Google Cloud TPU in the following weeks.
Sorry about that.
The project is in fast development and we are adding features on the PyTorch/XLA side, which need a counterpart on the Cloud TPU service side.

**Comment by williamFalcon** [comment]
ok thanks. Any way to build this now doing something else? just looking to train a resnet

**Comment by dlibenzi** [comment]
You can probably hack it up by making `ExecuteChained()` in `xrt_computation_client.cc` call directly the `ExecuteChainedSplit()` API and removing the `ExecuteChainedXrt()` API.
Also removing this line:

https://github.com/pytorch/xla/blob/ad330b336486c5a4c7d3665bbe749ddaf505b858/third_party/xla_client/xrt_computation_client.cc#L1012

And this function:

https://github.com/pytorch/xla/blob/ad330b336486c5a4c7d3665bbe749ddaf505b858/third_party/xla_client/xrt_computation_client.cc#L1077

**Comment by williamFalcon** [comment]
ok awesome. so make these changes and compile?

**Comment by qinjian623** [comment]
Hi,
I've hacked the code.
But another error happened here:

`
return slices[i].to(device=torch.device(self._devices[i]))
`

Here is the stack trace:
>    
	tensorflow::CurrentStackTrace[abi:cxx11]()
	xla::XrtComputationClient::TransferToServer(absl::Span<xla::ComputationClient::TensorSource const>)
	torch_xla::TensorToXlaData(at::Tensor const&, torch_xla::Device const&)
	
	torch_xla::XLATensor::GetIrValueForScalar(c10::Scalar, xla::PrimitiveType, torch_xla::Device const&)
	torch_xla::XLATensor::GetIrValueForScalar(c10::Scalar, xla::Shape const&, torch_xla::Device const&)
	torch_xla::XLATensor::full(absl::Span<long long const>, c10::Scalar, torch_xla::Device const&, c10::ScalarType)
	torch_xla::AtenXlaType::full(c10::ArrayRef<long>, c10::Scalar, c10::TensorOptions const&) const
	torch_xla::AtenXlaType::zeros(c10::ArrayRef<long>, c10::TensorOptions const&) const
	torch::autograd::VariableType::empty(c10::ArrayRef<long>, c10::TensorOptions const&) const
	at::native::to(at::Tensor const&, c10::TensorOptions const&, bool, bool)
	at::TypeDefault::to(at::Tensor const&, c10::TensorOptions const&, bool, bool) const
	torch::autograd::VariableType::to(at::Tensor const&, c10::TensorOptions const&, bool, bool) const
	
	
	_PyMethodDef_RawFastCallKeywords
	_PyCFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallDict
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	PyObject_Call
	
	
	
	clone

**Comment by dlibenzi** [comment]
Can you post the full error?

Are you trying to implement multi-core?
For that, should be using our `data_parallel` API. See example here:

https://github.com/pytorch/xla/blob/92c08e067404a6fa197386da59144a5644cdb288/test/test_train_mnist.py#L88

**Comment by dlibenzi** [comment]
> ok awesome. so make these changes and compile?

Yep

**Comment by williamFalcon** [comment]
ok... this failed for me...

Here are my modifications.
https://github.com/williamFalcon/xla/commit/24513927b8b6c7dee2000e9805409d893c1208aa

I assume there's something wrong  with my changes?

**Comment by dlibenzi** [comment]
> ok... this failed for me...
> 
> Here are my modifications.
> [williamFalcon@2451392](https://github.com/williamFalcon/xla/commit/24513927b8b6c7dee2000e9805409d893c1208aa)
> 
> I assume there's something wrong with my changes?

You have to remove the `ExecuteChainedXrt()` API from .cpp and .h

**Comment by williamFalcon** [comment]
duh... been about 4 years since I wrote c++, forgot about header files haha.

look ok now?
https://github.com/williamFalcon/xla/tree/master/third_party/xla_client

**Comment by dlibenzi** [comment]
> duh... been about 4 years since I wrote c++, forgot about header files haha.
> 
> look ok now?
> https://github.com/williamFalcon/xla/tree/master/third_party/xla_client

A build will tell faster than me going through the code :)

**Comment by williamFalcon** [comment]
Resolved. This error is gone. but there's a new error now (or just hangup...) Going to different issue.

my fork has the suggested hack in the meantime:
https://github.com/williamFalcon/xla

**Comment by qinjian623** [comment]
Hi, @williamFalcon 
Would you please tell me details of the new issue?
I've a new error when transfering data to devices.

**Comment by qinjian623** [comment]
@dlibenzi 
Hi, below is the full error.
Do I need to remove the cache before re-compling?
Or just re-run python setup.py install under pytorch/xla/?

Actually, this error happened when running python test/test_train_mnist.py.


```
======================================================================
ERROR: test_accurracy (__main__.TrainMnist)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_train_mnist.py", line 136, in test_accurracy
    self.assertGreaterEqual(train_mnist(), FLAGS.target_accuracy)
  File "test/test_train_mnist.py", line 88, in train_mnist
    model_parallel = dp.DataParallel(MNIST, device_ids=devices)
  File "/home/xx/pytorch/xla/torch_xla_py/data_parallel.py", line 162, in __init__
    module = network().to(device=torch.device(device))
  File "/home/xx/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 386, in to
    return self._apply(convert)
  File "/home/xx/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 193, in _apply
    module._apply(fn)
  File "/home/xx/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 199, in _apply
    param.data = fn(param.data)
  File "/home/xx/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 384, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:105 : Check failed: session_work.first->session()->Run( session_work.second.feed_inputs, session_work.second.outputs_handles, &outputs) == ::tensorflow::Status::OK() (Not found: Op type not registered 'XRTAllocateFromTensor' in binary running on n-a66176ae-w-0. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed. vs. OK)
*** Begin stack trace ***
	tensorflow::CurrentStackTrace[abi:cxx11]()
	xla::XrtComputationClient::TransferToServer(absl::Span<xla::ComputationClient::TensorSource const>)
	torch_xla::TensorToXlaData(at::Tensor const&, torch_xla::Device const&)
	
	torch_xla::XLATensor::GetIrValueForScalar(c10::Scalar, xla::PrimitiveType, torch_xla::Device const&)
	torch_xla::XLATensor::GetIrValueForScalar(c10::Scalar, xla::Shape const&, torch_xla::Device const&)
	torch_xla::XLATensor::full(absl::Span<long long const>, c10::Scalar, torch_xla::Device const&, c10::ScalarType)
	torch_xla::AtenXlaType::full(c10::ArrayRef<long>, c10::Scalar, c10::TensorOptions const&) const
	torch_xla::AtenXlaType::zeros(c10::ArrayRef<long>, c10::TensorOptions const&) const
	torch::autograd::VariableType::empty(c10::ArrayRef<long>, c10::TensorOptions const&) const
	at::native::to(at::Tensor const&, c10::TensorOptions const&, bool, bool)
	at::TypeDefault::to(at::Tensor const&, c10::TensorOptions const&, bool, bool) const
	torch::autograd::VariableType::to(at::Tensor const&, c10::TensorOptions const&, bool, bool) const
	
	
	_PyMethodDef_RawFastCallKeywords
	_PyMethodDescr_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	PyObject_Call
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	PyObject_Call
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	PyObject_Call
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyFunction_FastCallKeywords
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	_PyFunction_FastCallDict
	_PyObject_Call_Prepend
	
	_PyObject_FastCallKeywords
	_PyEval_EvalFrameDefault
	
	_PyEval_EvalFrameDefault
	_PyEval_EvalCodeWithName
	PyEval_EvalCodeEx
	PyEval_EvalCode
	
	PyRun_FileExFlags
	PyRun_SimpleFileExFlags
	
	_Py_UnixMain
	__libc_start_main
	
*** End stack trace ***


----------------------------------------------------------------------
Ran 1 test in 0.163s

FAILED (errors=1)
```

**Comment by dlibenzi** [comment]
This is the error:

```
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:105 : Check failed: session_work.first->session()->Run( session_work.second.feed_inputs, session_work.second.outputs_handles, &outputs) == ::tensorflow::Status::OK() (Not found: Op type not registered 'XRTAllocateFromTensor' in binary running on n-a66176ae-w-0. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed. vs. OK)
```

Same issue of williamFalcon. You are using a TPU VM version which is too old.
As I said to williamFalcon a new version will be out soon, and we are working on having Cloud TPU users to be able to select the TPU VM nightly builds.
Sorry about that, and thanks for you patience.
